libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/VoiceInstructionLoader.java

```diff
@@ -0,0 +1,198 @@
+import com.mapbox.api.speech.v1.MapboxSpeech;
+import com.mapbox.services.android.navigation.v5.routeprogress.RouteProgress;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import okhttp3.Cache;
+import okhttp3.ResponseBody;
+import retrofit2.Call;
+import retrofit2.Callback;
+import retrofit2.Response;
+
+@AutoValue
+public abstract class VoiceInstructionLoader {
+  private static final int NUMBER_TO_CACHE = 3;
+  private static VoiceInstructionLoader instance = null;
+
+  /**
+   * Returns the singleton instance of VoiceInstructionLoader. It must first be initialized through
+   * a builder.
+   *
+   * @return current instance if it's initialized, or null
+   */
+  public static synchronized VoiceInstructionLoader getInstance() {
+    return instance;
+  }
+
+  /**
+   * Makes the call to MapboxSpeech to get the given string instruction as a sound file.
+   *
+   * @param instruction text to dictate
+   * @param textType "ssml" or "text"
+   * @param callback to relay retrofit status
+   */
+  public void getInstruction(String instruction, String textType, Callback<ResponseBody> callback) {
+    getMapboxBuilder()
+      .instruction(instruction)
+      .textType(textType)
+      .build()
+      .enqueueCall(callback);
+  }
+
+  /**
+   * Makes call to MapboxSpeech with empty callbacks. This is so that the result is cached in the
+   * cache specified in the builder.
+   *
+   * @param routeProgress to get instructions from
+   * @param isFirst whether this is the first call. This way, if we're caching three ahead, on the
+   *                first call the first two instructions will also be cached.
+   */
+  public void cacheInstructions(RouteProgress routeProgress, boolean isFirst) {
+    List<VoiceInstructions> voiceInstructions = getNextInstructions(routeProgress);
+
+    if (isFirst) {
+      for (int i = 0; i <= NUMBER_TO_CACHE - 1; i++) {
+        if (voiceInstructions.size() > i) {
+          cacheInstruction(voiceInstructions.get(i).ssmlAnnouncement());
+        }
+      }
+    } else {
+      if (voiceInstructions.size() >= NUMBER_TO_CACHE) {
+        cacheInstruction(voiceInstructions.get(NUMBER_TO_CACHE - 1).ssmlAnnouncement());
+      }
+    }
+  }
+
+  private void cacheInstruction(String instruction) {
+    getMapboxBuilder().instruction(instruction).build().enqueueCall(new Callback<ResponseBody>() {
+      @Override
+      public void onResponse(Call<ResponseBody> call, Response<ResponseBody> response) {
+
+      }
+
+      @Override
+      public void onFailure(Call<ResponseBody> call, Throwable throwable) {
+
+      }
+    });
+  }
+
+  @Nullable
+  abstract String language();
+
+  @Nullable
+  abstract String textType();
+
+  @Nullable
+  abstract String outputType();
+
+  @Nullable
+  abstract Cache cache();
+
+  @NonNull
+  abstract String accessToken();
+
+  private MapboxSpeech.Builder getMapboxBuilder() {
+    MapboxSpeech.Builder builder = MapboxSpeech.builder().accessToken(accessToken());
+
```

Comment:
NIT empty lines 👇

Reply:
I added those for readability but I can remove them